src/main/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheck.java

```diff
@@ -0,0 +1,356 @@
+    private static final int DEFAULT_MAX_COMPONENTS = 8;
+
+    /** Specify the maximum number of {@code private} components allowed. */
+    private int maxPrivate = DEFAULT_MAX_COMPONENTS;
+    /** Specify the maximum number of {@code package} components allowed. */
+    private int maxPackage = DEFAULT_MAX_COMPONENTS;
+    /** Specify the maximum number of {@code protected} components allowed. */
+    private int maxProtected = DEFAULT_MAX_COMPONENTS;
+    /** Specify the maximum number of {@code public} components allowed. */
+    private int maxPublic = DEFAULT_MAX_COMPONENTS;
+
+    /**
+     * Setter to specify the maximum number of {@code private} components allowed.
+     *
+     * @param value the maximum allowed.
+     */
+    public void setMaxPrivate(int value) {
+        maxPrivate = value;
+    }
+
+    /**
+     * Setter to specify the maximum number of {@code package} components allowed.
+     *
+     * @param value the maximum allowed.
+     */
+    public void setMaxPackage(int value) {
+        maxPackage = value;
+    }
+
+    /**
+     * Setter to specify the maximum number of {@code protected} components allowed.
+     *
+     * @param value the maximum allowed.
+     */
+    public void setMaxProtected(int value) {
+        maxProtected = value;
+    }
+
+    /**
+     * Setter to specify the maximum number of {@code public} components allowed.
+     *
+     * @param value the maximum allowed.
+     */
+    public void setMaxPublic(int value) {
+        maxPublic = value;
+    }
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getAcceptableTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return new int[] {
+            TokenTypes.RECORD_DEF,
+        };
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.RECORD_DEF,
+        };
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        // Instantiate new component counter
+        final ComponentCounter counter = new ComponentCounter();
+
+        // Count record components in this record definition
+        countComponents(counter, ast);
+
+        // Check counter and report violations
+        checkCounter(counter, ast);
+    }
+
+    /**
+     * Determine the visibility modifier, then iterate through the record component list
+     * and increment the counter.
+     *
+     * @param counter the ComponentCounter instance for this check
+     * @param ast the ast to check
+     *
+     */
+    private static void countComponents(ComponentCounter counter, DetailAST ast) {
+        final DetailAST recordComponents =
+                ast.findFirstToken(TokenTypes.RECORD_COMPONENTS);
+        final DetailAST modifiers =
+                ast.findFirstToken(TokenTypes.MODIFIERS);
+
+        counter.setScope(ScopeUtil.getScopeFromMods(modifiers));
+
+        DetailAST componentToCount =
```

Comment:
This can be done using `TokenUtils::forEachChild`

Reply:
Can you give me an example?  `forEachChild()` performs an action on each child, I'm not sure how that applies here.

Reply:
```java
        TokenUtil.forEachChild(recordComponents,
                TokenTypes.RECORD_COMPONENT_DEF, node -> counter.increment());
```

What your code actually does:
1. For each child from the first child find one with type==TokenTypes.RECORD_COMPONENT_DEF.
2. Double check the type. Then increment the counter.
3. loop over the rest of children
It can be simplified to
```java
        DetailAST componentToCount = recordComponents.getFirstChild();
        while (componentToCount != null) {
            if (componentToCount.getType() == TokenTypes.RECORD_COMPONENT_DEF) {
                counter.increment();
            }
            componentToCount = componentToCount.getNextSibling();
        }
```
... and this is exactly what `TokenUtils::forEachChild` does:
1. loop for each child
2. compare type
3. if match, invoke callback

Reply:
Thanks a lot, @pbludov .  It is done.